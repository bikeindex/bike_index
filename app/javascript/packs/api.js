/*
  Private
*/

const url = urn => [process.env.BASE_URL, urn].join("/");

const serialSearchUrl = serial =>
  url(`api/v2/bikes_search?serial=${serial}`);

const fuzzySearchUrl = serial =>
  url(`api/v2/bikes_search/close_serials?serial=${serial}`);

const partialMatchSerialSearchUrl = params => {
  const query = queryString(params)
  return url(`api/v3/search/serials_containing?${query}`);
}

const serialCloseSearchUrl = params => {
  const query = queryString(params)
  return url(`api/v3/search/close_serials?${query}`);
}

const serialExternalSearchUrl = ({ serial }) => {
  const query = queryString({ serial })
  return url(`api/v3/search/external_registries?${query}`);
}

const request = async url => {
  const resp = await fetch(url);
  const json = await resp.json();
  return json;
};

const queryString = ({ serial, stolenness, location, query, query_items }) => {
  const params = {};
  if (serial) { params.serial = serial; }
  if (stolenness) { params.stolenness = stolenness; }
  if (location) { params.location = location; }
  if (query) { params.query = query; }
  if (query_items) { params.query_items = query_items; }

  return Object.keys(params).map(k => `${k}=${params[k]}`).join("&");
}

/*
  Public
*/

const fetchSerialResults = serial => {
  const url = serialSearchUrl(serial);
  return request(url);
};

const fetchFuzzyResults = serial => {
  const url = fuzzySearchUrl(serial);
  return request(url);
};

const fetchPartialMatchSearch = interpretedParams => {
  const url = partialMatchSerialSearchUrl(interpretedParams);
  return request(url);
}

const fetchSerialCloseSearch = interpretedParams => {
  const url = serialCloseSearchUrl(interpretedParams);
  return request(url);
}

const fetchSerialExternalSearch = ({ raw_serial }) => {
  const url = serialExternalSearchUrl({ serial: raw_serial });
  return request(url);
}

export default {
  fetchSerialResults,
  fetchFuzzyResults,
  fetchSerialExternalSearch,
  fetchSerialCloseSearch,
  fetchPartialMatchSearch,
};
