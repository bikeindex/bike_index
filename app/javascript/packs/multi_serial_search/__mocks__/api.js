const fetchSerialResults = async () => new Promise(resolve => resolve({}));

export {
  fetchSerialResults,
};
