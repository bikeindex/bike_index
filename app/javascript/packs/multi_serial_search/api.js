/* TODO: put some of the private implementation into a utility helper */

/*
  Private
*/

const isDevelopment = process.env.RAILS_ENV === 'development';

const serialSearchUrl = serial => {
  const urn = `api/v1/bikes?multi_serial_search=true&serial=${serial}`;
  return isDevelopment
    ? `http://localhost:3001/${urn}`
    : `https://bikeindex.org/${urn}`;
};

const request = async url => {
  const resp = await fetch(url);
  const json = await resp.json();
  return json;
};

/*
  Public
*/

const fetchSerialResults = async serial => {
  const url = serialSearchUrl(serial);
  const results = await request(url);
  return results;
};

export { fetchSerialResults };
