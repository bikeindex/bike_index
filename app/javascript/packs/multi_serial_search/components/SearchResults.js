import React from "react";
import PropTypes from "prop-types";
import Loading from "../../Loading";
import BikeList from "./BikeList";

const SearchResults = ({
  serialResults,
  loading,
  onFuzzySearch,
  fuzzySearching
}) =>
  loading ? (
    <Loading />
  ) : (
    serialResults && (
      <section id="ms_search_section">
        <span className="padded" />
        <div className="multiserial-response">
          <h2>Multi serial search results</h2>

          {/* Serial search chips */}
          <ul id="serials_submitted" className="multiserials-list">
            {serialResults.map(({ serial, anchor, bikes, fuzzyBikes }) => {
              const hasResults = bikes.length > 0 || fuzzyBikes.length > 0;
              return (
                <li
                  key={serial}
                  className={hasResults ? "ms-match" : "ms-nomatch"}
                  name={serial}
                >
                  {serial}
                  {hasResults && <a href={anchor} className="scroll-to-ref" />}
                </li>
              );
            })}
          </ul>

          {/* Fuzzy search button  */}
          {!fuzzySearching && (
            <div className="multiserial-fuzzy-box">
              <button
                type="submit"
                className="btn-primary"
                onClick={onFuzzySearch}
              >
                Include closely matching serials in results
              </button>
            </div>
          )}

          {/* Bike Results */}
          <div id="bikes_returned">
            {serialResults.map(({ bikes, fuzzyBikes, serial, anchor }) => {
              if (bikes.length === 0 && fuzzyBikes.length === 0) return;
              const id = anchor.slice(1);
              return (
                <div key={id} id={id} className="multiserial-results">
                  <BikeList serial={serial} bikes={bikes} />

                  {fuzzySearching && (
                    <BikeList
                      serial={serial}
                      bikes={fuzzyBikes}
                      fuzzySearching={fuzzySearching}
                    />
                  )}
                </div>
              );
            })}
          </div>
        </div>
      </section>
    )
  );

SearchResults.propTypes = {
  serialResults: PropTypes.arrayOf(PropTypes.object),
  loading: PropTypes.bool,
  fuzzySearching: PropTypes.bool,
  onFuzzySearch: PropTypes.func
};

export default SearchResults;
