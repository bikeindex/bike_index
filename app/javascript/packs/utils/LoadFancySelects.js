import selectize from "selectize";

// Right now we're using selectize.js (https://selectize.github.io/selectize.js/) to make our select boxes fancy.
// This could be changed to some other library at some point - this makes it possible for us to abstract that away.
// Just add the `unfancy` and `fancy-select` classes to a select box and it will be a fancy select box!
const LoadFancySelects = () => {
  // Gross, check if
  $(".unfancy.fancy-select.no-restore-on-backspace select").selectize({
    create: false,
    plugins: []
  });

  $(".unfancy.fancy-select select").selectize({
    create: false,
    plugins: ["restore_on_backspace"]
  });
  // Remove them so we don't initialize twice
  $(".unfancy.fancy-select, .unfancy.fancy-select-placeholder").removeClass(
    "unfancy"
  );
};

export default LoadFancySelects;
