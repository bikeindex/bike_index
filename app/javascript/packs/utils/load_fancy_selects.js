import selectize from "selectize";

// Right now we're using selectize.js (https://selectize.github.io/selectize.js/) to make our select boxes fancy.
// This could be changed to some other library at some point - this makes it possible for us to abstract that away.
// Just add the `unfancy` and `fancy-select` classes to a select box and it will be a fancy select box!
const LoadFancySelects = () => {
  $(".unfancy.fancy-select.no-restore-on-backspace select").selectize({
    create: false,
    plugins: []
  });
  // Remove, because we've already added them
  $(".unfancy.fancy-select.no-restore-on-backspace").removeClass("unfancy");

  $(".unfancy.fancy-select select").selectize({
    create: false,
    plugins: ["restore_on_backspace"]
  });
  // Remove them so we don't initialize twice
  $(".unfancy.fancy-select").removeClass("unfancy");
};

export default LoadFancySelects;
