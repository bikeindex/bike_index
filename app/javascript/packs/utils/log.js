import loglevel from "loglevel";

if (process.env.RAILS_ENV !== "production") loglevel.setLevel("debug");

export default loglevel;
