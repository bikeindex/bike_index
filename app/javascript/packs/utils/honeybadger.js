import Honeybadger from "honeybadger-js";
import log from "./log";

const environment = process.env.RAILS_ENV;

/* webpacker can't access this exported env */
const apiKey =
  environment === "production" ? process.env.HONEYBADGER_FRONTEND_API_KEY : "foo";

const config = {
  apiKey,
  environment: process.env.RAILS_ENV,
  disabled: environment !== "production"
};
const honeybadger = Honeybadger.configure(config);

Honeybadger.beforeNotify(notice => {
  if (environment === "development") {
    return log.debug("Error", notice);
  }
});
export default honeybadger;
