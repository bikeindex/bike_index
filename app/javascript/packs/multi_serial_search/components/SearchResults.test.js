/* eslint import/no-unresolved: 0 */

import React from "react";
import { shallow, mount } from "enzyme";
import toJson from "enzyme-to-json";
import { bikeResponse } from "helpers/bikes";
import { event } from "helpers/utils";
import api from "../../api";
import SearchResults from "./SearchResults";

/*
  api mocks
*/
jest.mock("../../api");
api.fetchFuzzyResults.mockImplementation(
  () => new Promise(resolve => resolve(bikeResponse))
);

/*
  helpers
*/
const noResults = [{ serial: "one", bikes: [], fuzzyBikes: [] }];
const serialResults = [
  {
    ...bikeResponse,
    fuzzyBikes: [],
    serial: "two",
    anchor: "#mock"
  }
];

describe("<SearchResults />", () => {
  it("matches the snapshot", () => {
    const tree = shallow(<SearchResults />);
    expect(toJson(tree)).toMatchSnapshot();
  });

  it("has a loading indicator when loading", () => {
    const wrap = mount(<SearchResults loading />);
    expect(wrap.exists("Loading")).toBe(true);
  });

  it("is not loading after fetching results", () => {
    const wrap = mount(
      <SearchResults loading={false} serialResults={noResults} />
    );
    expect(wrap.exists("Loading")).toBe(false);
  });

  describe("serialResults", () => {
    it("has a no-match class as default", () => {
      const wrap = mount(
        <SearchResults loading={false} serialResults={noResults} />
      );
      expect(wrap.exists("li.ms-nomatch")).toBe(true);
    });

    it("has a match class if there is a match", () => {
      const wrap = mount(
        <SearchResults loading={false} serialResults={serialResults} />
      );
      expect(wrap.exists("li.ms-nomatch")).toBe(false);
      expect(wrap.exists("li.ms-match")).toBe(true);
    });

    it("has a div id with the normalized anchor", () => {
      const wrap = mount(
        <SearchResults loading={false} serialResults={serialResults} />
      );
      const anchorId = serialResults[0].anchor.slice(1);
      expect(wrap.find(".multiserial-results").prop("id")).toEqual(anchorId);
    });

    it("has a BikeList for the matching serial", () => {
      const wrap = mount(
        <SearchResults loading={false} serialResults={serialResults} />
      );
      const result = serialResults[0];
      expect(wrap.find("BikeList").prop("serial")).toEqual(result.serial);
    });
  });

  it("calls onFuzzySearch on button click", () => {
    const mockFuzzy = jest.fn(() => "mock api request");
    const wrap = mount(
      <SearchResults
        loading={false}
        onFuzzySearch={mockFuzzy}
        serialResults={serialResults}
      />
    );
    wrap.find("button").simulate("click", event);
    expect(mockFuzzy).toHaveBeenCalledTimes(1);
  });
});
