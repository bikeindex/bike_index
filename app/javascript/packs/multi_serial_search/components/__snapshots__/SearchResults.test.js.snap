// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchResults /> matches the snapshot 1`] = `""`;
